src/ppx/scene/scene_mesh.cpp

```diff
@@ -0,0 +1,124 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/scene/scene_mesh.h"
+#include "ppx/grfx/grfx_device.h"
+
+namespace ppx {
+namespace scene {
+
+// -------------------------------------------------------------------------------------------------
+// MeshData
+// -------------------------------------------------------------------------------------------------
+MeshData::MeshData(
+    const scene::VertexAttributeFlags& availableVertexAttributes,
+    grfx::Mesh*                        pGpuMesh)
+    : mAvailableVertexAttributes(availableVertexAttributes), mGpuMesh(pGpuMesh)
+{
+    PPX_ASSERT_MSG(!IsNull(mGpuMesh.Get()), "mGpuMesh is NULL");
+
+    mIndexBufferView.pBuffer   = mGpuMesh->GetIndexBuffer();
+    mIndexBufferView.indexType = mGpuMesh->GetIndexType();
+    mIndexBufferView.offset    = 0;
+    mIndexBufferView.size      = mGpuMesh->GetIndexBuffer()->GetSize();
+
+    const uint32_t kPositionIndex = 0;
+    mPositionBufferView.pBuffer   = mGpuMesh->GetVertexBuffer(kPositionIndex);
+    mPositionBufferView.stride    = mGpuMesh->GetDerivedVertexBindings()[kPositionIndex].GetStride();
+    mPositionBufferView.offset    = 0;
+    mPositionBufferView.size      = mGpuMesh->GetVertexBuffer(kPositionIndex)->GetSize();
+
+    if (mAvailableVertexAttributes.mask != 0) {
+        const uint32_t kAttributesIndex = 1;
+        mAttributeBufferView.pBuffer    = mGpuMesh->GetVertexBuffer(kAttributesIndex);
+        mAttributeBufferView.stride     = mGpuMesh->GetDerivedVertexBindings()[kAttributesIndex].GetStride();
+        mAttributeBufferView.offset     = 0;
+        mAttributeBufferView.size       = mGpuMesh->GetVertexBuffer(kAttributesIndex)->GetSize();
+    }
+}
+
+MeshData::~MeshData()
+{
+    if (mGpuMesh) {
+        auto pDevice = mGpuMesh->GetDevice();
+        pDevice->DestroyMesh(mGpuMesh);
+        mGpuMesh.Reset();
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+// Mesh
+// -------------------------------------------------------------------------------------------------
+Mesh::Mesh(
+    const scene::MeshDataRef&            meshData,
+    std::vector<scene::PrimitiveBatch>&& batches)
+    : mMeshData(meshData),
+      mBatches(std::move(batches))
+{
+    UpdateBoundingBox();
+}
+
+Mesh::Mesh(
+    std::unique_ptr<scene::ResourceManager>&& resourceManager,
+    const scene::MeshDataRef&                 meshData,
+    std::vector<scene::PrimitiveBatch>&&      batches)
+    : mResourceManager(std::move(resourceManager)),
+      mMeshData(meshData),
+      mBatches(std::move(batches))
+{
+    UpdateBoundingBox();
+}
+
+Mesh::~Mesh()
+{
+    if (mResourceManager) {
```

Comment:
a bit lost here, why `mResourceManager->DestroyAll()` in destructor of Mesh class? isn't `ResourceManager` contains resources of all meshes? Does this mesh class contains all meshes in the scene?

Reply:
It's explained in the comments that when a mesh is load standalone it uses a local resource manager to store the geometry data that's shared across the primitives.